java/src/main/java/com/logicalclocks/hsfs/engine/Utils.java

```diff
@@ -48,10 +49,22 @@
     return features;
   }
 
-  // TODO(Fabio): keep into account the sorting - needs fixing in Hopsworks as well
-  public void schemaMatches(Dataset<Row> dataset, List<Feature> features) throws FeatureStoreException {
-    StructType tdStructType = new StructType(features.stream().map(
-        f -> new StructField(f.getName(),
+  public List<TrainingDatasetFeature> parseTrainingDatasetSchema(Dataset<Row> dataset) throws FeatureStoreException {
+    List<TrainingDatasetFeature> features = new ArrayList<>();
+
+    int index = 0;
+    for (StructField structField : dataset.schema().fields()) {
+      // TODO(Fabio): unit test this one for complext types
+      features.add(new TrainingDatasetFeature(structField.name(), structField.dataType().catalogString(), index++));
+    }
+
+    return features;
+  }
+
+  public void schemaMatches(Dataset<Row> dataset, List<TrainingDatasetFeature> features) throws FeatureStoreException {
```

Comment:
maybe we can call it `trainingDatasetSchemaMatches` to avoid confusion
```suggestion
  public void trainingDatasetSchemaMatches(Dataset<Row> dataset, List<TrainingDatasetFeature> features) throws FeatureStoreException {
```

Reply:
Ok